provider/formtype_test.go

```diff
@@ -0,0 +1,426 @@
+				optionType:    provider.OptionTypeListString,
+				customOptions: []string{"red", "blue", "green"},
+				defValue:      `["red", "yellow"]`,
+				styling:       nil,
+			},
+			expectError: regexp.MustCompile("is not a valid option"),
+		},
+	}
+
+	passed := t.Run("TabledTests", func(t *testing.T) {
+		// TabledCases runs through all the manual test cases
+		for _, c := range cases {
+			t.Run(c.name, func(t *testing.T) {
+				t.Parallel()
+				if _, ok := formTypesChecked[c.config.String()]; ok {
+					t.Log("Duplicated form type check, delete this extra test case")
+					t.Fatalf("form type %q already checked", c.config.String())
+				}
+
+				formTypesChecked[c.config.String()] = struct{}{}
+				formTypeTest(t, c)
+			})
+		}
+	})
+
+	if !passed {
+		// Do not run additional tests and pollute the output
+		t.Log("Tests failed, will not run the assumed error cases")
+		return
+	}
+
+	// AssumeErrorCases assumes any uncovered test will return an error. Not covered
+	// cases in the truth table are assumed to be invalid. So if the tests above
+	// cover all valid cases, this asserts all the invalid cases.
+	//
+	// This test consequentially ensures all valid cases are covered manually above.
+	t.Run("AssumeErrorCases", func(t *testing.T) {
+		// requiredChecks loops through all possible form_type and option_type
+		// combinations.
+		requiredChecks := make([]formTypeCheck, 0)
+		for _, ft := range append(provider.ParameterFormTypes(), "") {
+			for _, ot := range provider.OptionTypes() {
+				requiredChecks = append(requiredChecks, formTypeCheck{
+					formType:   ft,
+					optionType: ot,
+					options:    false,
+				})
+				requiredChecks = append(requiredChecks, formTypeCheck{
+					formType:   ft,
+					optionType: ot,
+					options:    true,
+				})
+			}
+		}
+
+		for _, check := range requiredChecks {
+			_, alreadyChecked := formTypesChecked[check.String()]
+			if alreadyChecked {
+				continue
+			}
+
+			ftName := check.formType
+			if ftName == "" {
+				ftName = "default"
+			}
+			fc := formTypeTestCase{
+				name: fmt.Sprintf("%s_%s_%t",
+					ftName,
+					check.optionType,
+					check.options,
+				),
+				config:      check,
+				assert:      paramAssert{},
+				expectError: regexp.MustCompile("is not supported"),
+			}
+
+			t.Run(fc.name, func(t *testing.T) {
+				t.Parallel()
+
+				// This is just helpful log output to give the boilerplate
+				// to write the manual test.
+				tcText := fmt.Sprintf(`
+					obvious(%s, ezconfigOpts{
+						Options:    %t,
+						OptionType: %q,
+						FormType:   %q,
+					}),
+				//`, "<expected_form_type>", check.options, check.optionType, check.formType)
+
+				probablyPassed := formTypeTest(t, fc)
+				if !probablyPassed {
+					t.Logf("To construct this test case:\n%s", tcText)
+				}
+			})
+
+		}
+	})
+}
+
+// ezconfig converts a formTypeCheck into a terraform config string.
+func ezconfig(paramName string, cfg formTypeCheck) (defaultValue string, tf string) {
```

Comment:
nit: rename for clarity, `ezconfig` isn't particularly illuminating as an identifier.